ls1intum/Artemis
# Feature/reduce websocket subscriptions

### Checklist

### Motivation and Context
Currently, a new WebSocket subscription is opened for every unrated exercise where the due date passed and for every started programming exercise for every student in the course overview. Also, for TAs and higher, a WebSocket subscription is opened for every student once they view an exercise overview.

In large courses like EIST, around 1500 students use the system at the same time. Every one of these students typically has quite some unrated exercises and started programming exercises, sometimes leading to > 10 subscriptions per students. The subscribe-mechanism is quite exhausting for the server which already lead to problems during one quiz in a live lecture as there was a large queue with unprocessed subscriptions, decreasing the user experience and functionality as other WebSocket messages could also not be processed.

This PR creates one subscription for all `/newResults`-messages and one for every `/newSubmission`-messages for each student. Also, TAs and higher will now open one subscription per exercise, instead of one per participation.

### Description
A new topic `/user/topic/newResults` is introduced which every client can independently subscribe to. The server will then send the new results to this topic for the student(s) owning the participation instead of a topic only used by this participation.
The new result will also be broadcasted to the topic `/topic/exercise/{exerciseId}/newResults`, where TAs and higher can subscribe to.
A security check was added in the server that ensures that only TAs and higher can subscribe to the `/topic/exercise/{exerciseId}/newResults` topic.

In the client, at several occurrences, it has to be specified if a subscription to a new result for a participation is for a student or for an instructor. This is done by adding a property `personalParticipation` to several components. This value will be set be their parent components, depending on the context the child component is used (e.g. in the course overview for a student, the `UpdatingResultComponent` will be used in a personal context whereas it will be used in a non-personal context in the exercise view for instructors (where the state of the template & solution repositories are shown)).

The client also keeps a map which stores which topic was used by which participation, therefore ensuring that every topic will only be subscribed once.
This logic also ensures that an unsubscription from one specific WebSocket topic will only occur if no other participations rely on this topic.

#### The difference between the previous subscriptions and the current subscriptions can be found in the following tables:
| Old subscriptions            | Student                                              | TA/Instructor/Admin                                  |
|------------------------------|------------------------------------------------------|------------------------------------------------------|
| Inform about new results     | /topic/participation/{participationId}/newResults    | /topic/participation/{participationId}/newResults    |
| Inform about new submissions | /topic/participation/ {particpationId}/newSubmission | /topic/participation/ {particpationId}/newSubmission |

| New subscriptions            | Student                   | TA/Instructor/Admin                        |
|------------------------------|---------------------------|--------------------------------------------|
| Inform about new results     | /user/topic/newResults    | /topic/exercise/{exerciseId}/newResults    |
| Inform about new submissions | /user/topic/newSubmission | /topic/exercise/{exerciseId}/newSubmission |

The number websocket subscriptions is reduced as 2 subscriptions per student are enough and also the instructors only need 2 subscriptions per exercise instead of per participation.

### Steps for Testing 
In general, results and subscriptions should be updated without reload wherever they are updated before this PR, but in a more performant way.

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a text exercise
4. Participate as a student
5. Assess the submission of the student
6. The student should be informed without reload about the new result

![Peek 2020-06-04 11-46](https://user-images.githubusercontent.com/5084100/83742047-3b92d480-a659-11ea-88b6-781eae5d5bb0.gif)

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise
4. Click on the View-button for the exercise.
5. Scroll down, the status of the template and solution participation should be displayed.
6. Re-trigger the builds for template and solution participations. Only the participation that you triggered should be changed to "Building and testing..." and the received result should be displayed. (**NOTE**: this can take some time as the build is processed in the background on the CI-server, but it should normally take not significantly longer than it took for me locally (see GIF))
![Peek 2020-06-04 12-01](https://user-images.githubusercontent.com/5084100/83743800-831a6000-a65b-11ea-94f5-6411c4947cc1.gif)

7. Start the programming exercise as a student
8. Change some code and submit, the text should change to "Testing and Building..."
9. After the build was executed, the result should be displayed.

![Peek 2020-06-04 15-10](https://user-images.githubusercontent.com/5084100/83760779-b79b1580-a675-11ea-853e-a5df1f7c95cf.gif)
